### PR TITLE
feat(dl_tuning_spaces): Add tuning spaces from Gorishniy 2021 "Revisiting" article

### PR DESCRIPTION
**Pull request by cxzhang4**
As requested by Sebastian. 

Current state is not ready to be merged yet - we need to discuss modifications, like how to incorporate early stopping.

### CODE REVIEW

File: R/tuning_spaces_dl.R [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#' @title Deep Learning Tuning Spaces
+#'
+#' @name mlr_tuning_spaces_dl
+#'
+#' @description
+#' Tuning spaces for deep neural network architectures from the `r cite_bib("gorishniy2021revisting")` article.
+#'
```
Comment (sebffischer) [inline R/tuning_spaces_dl.R]: Requires an optimizer with a `weight_decay` parameter, such as AdamW

File: R/tuning_spaces_dl.R [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#' When the article suggests multiple ranges for a given hyperparameter, these tuning spaces choose the widest range.
+#'
+#' @source
+#' `r format_bib("gorishniy2021revisting")`
+#'
+#' @aliases
+#' mlr_tuning_spaces_classif.ft_transformer.default
+#'
+#' @section FT-Transformer tuning space
+#' `r rd_info(lts("classif.ft_transformer.default"))`
```
Comment (sebffischer) [inline R/tuning_spaces_dl.R]:
```suggestion
#' `r rd_info(lts("classif.ft_transformer.rtdl"))`
```

File: R/tuning_spaces_dl.R [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  epochs = 100
+)
+
+add_tuning_space(
+  id = "classif.mlp.default",
+  values = vals,
+  tags = c("gorishniy2021", "classification"),
+  learner = "classif.mlp",
+  package = "mlr3torch",
+  label = "Classification MLP with \"Revisiting\""
```
Comment (sebffischer) [inline R/tuning_spaces_dl.R]:
```suggestion
  label = "Classification MLP with RTDL"
```

File: R/tuning_spaces_dl.R [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+# resnet
+vals = list(
+  n_blocks = to_tune(1, 16),
+  d_block = to_tune(64, 1024),
+  d_hidden_multiplier = to_tune(1, 4),
+  dropout1 = to_tune(0, 0.5),
+  dropout2 = to_tune(0, 0.5),
+  opt.lr = to_tune(1e-5, 1e-2, logscale = TRUE),
+  opt.weight_decay = to_tune(1e-6, 1e-3, logscale = TRUE),
+  epochs = 100
```
Comment (sebffischer) [inline R/tuning_spaces_dl.R]:
```suggestion
  epochs = to_tune(upper = 100, tags = "internal_tuning")
```

File: R/tuning_spaces_dl.R (hunk 1 above)
Comment (sebffischer) [reply R/tuning_spaces_dl.R]: Mention that we use early stopping and that the user has to configure the `patience` and `measures_valid` parameters.

File: R/tuning_spaces_rtdl.R [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+# mlp
+vals = c(
+  n_layers = to_tune(1, 16),
+  neurons = to_tune(1, 1024),
+  p = to_tune(0, 0.5),
+  opt.lr = to_tune(1e-5, 1e-2, logscale = TRUE),
+  opt.weight_decay = to_tune(1e-6, 1e-3, logscale = TRUE),
+  epochs = to_tune(upper = 100L, internal = TRUE),
+  patience = 17
```
Comment (cxzhang4) [inline R/tuning_spaces_rtdl.R]: Relevant quote from section 4.3 of the paper:

"We continue training until there are patience + 1
consecutive epochs without improvements on the validation set; we set patience = 16 for all
algorithms."

Our `patience` parameter stops when there are `patience` epochs without improvement, so using 17 here is equivalent to their 16.

**Review by cxzhang4 — COMMENTED** [review]
Added `patience` based on the paper's description of their tuning process

File: R/tuning_spaces_rtdl.R [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+#' @include mlr_tuning_spaces.R
+NULL
+
+# mlp
+vals = c(
+  n_layers = to_tune(1, 16),
+  neurons = to_tune(1, 1024),
+  p = to_tune(0, 0.5),
+  opt.lr = to_tune(1e-5, 1e-2, logscale = TRUE),
+  opt.weight_decay = to_tune(1e-6, 1e-3, logscale = TRUE),
```
Comment (cxzhang4) [inline R/tuning_spaces_rtdl.R]: TODO: investigate further and decide how to set `opt.weight_decay` in light of the following quote:

"We follow the popular “transformers” library (Wolf et al., 2020) and do not apply weight decay to Feature Tokenizer, biases in linear layers and normalization layers."

Comment (cxzhang4) [reply R/tuning_spaces_rtdl.R]: Do we want to handle this? We could try making default parameter groups for the optimizers. [Here's a Python example from the `transformers` library](https://github.com/huggingface/transformers/blob/9cd7570f34fdb833ed874b2eba4d4ea3ae9ccb03/examples/legacy/run_openai_gpt.py#L229C9-L237C114).

Comment (sebffischer) [reply R/tuning_spaces_rtdl.R]: `param_groups = function(network) { ... # disable the weight decay for some parameters}`

File: R/tuning_spaces_rtdl.R [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+add_tuning_space(
+  id = "regr.tab_resnet.rtdl",
+  values = vals,
+  tags = c("gorishniy2021", "regression"),
+  learner = "regr.tab_resnet",
+  package = "mlr3torch",
+  label = "Regression Tabular ResNet with RTDL"
+)
+
+no_wd = function(name) {
```
Comment (cxzhang4) [inline R/tuning_spaces_rtdl.R:103]: They have another implementation that removes weight decay for all bias layers in https://github.com/yandex-research/rtdl-revisiting-models/blob/main/package/rtdl_revisiting_models.py, and I provide an implementation below.

This does not match the paper's description, which says they remove weight decay only for biases in linear layers.

```
  other_no_wd_params = c("embedding", "_normalization", ".bias")

  return(any(map_lgl(other_no_wd_params, function(pattern) grepl(pattern, name, fixed = TRUE))))
```

File: R/tuning_spaces_rtdl.R [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+  opt.lr = to_tune(1e-5, 1e-3, logscale = TRUE),
+  opt.weight_decay = to_tune(1e-6, 1e-3, logscale = TRUE),
+  opt.param_groups = to_tune(list(rtdl_param_groups = rtdl_param_groups)),
+  epochs = to_tune(upper = 100L, internal = TRUE),
+  patience = 17L
+)
+
+vals = list(
+  n_blocks = to_tune(1, 6),
+  d_token = to_tune(p_int(8, 64, trafo = function(x) 8 * x)),
```
Comment (cxzhang4) [inline R/tuning_spaces_rtdl.R]: Constructs the paper's tuning space for `d_token` (a uniform sequence of integers from 64 to 512) while ensuring that `d_token` is divisible by `attention_n_heads`, which has an initial value of 8.

**Comment by cxzhang4** [comment]
Moved to [a branch](https://github.com/mlr-org/mlr3tuningspaces/pull/61)
